File: test-api-pkg/ts/test-docs-databoard.ts
/**
 * Mirrors website/public/sdk/databoard.md against @imbrace/sdk@1.0.4 (npm)
 * — API-key auth.
 *
 * Sections: Board CRUD · Items · Search · Fields/Segments/Export.
 *
 * Doc-gaps the test calls out:
 *   - `boards.createItem` doc shape  `{ fields: { name, status, value } }`
 *     fails 400; canonical is `{ fields: [{ board_field_id, value }] }`.
 *   - `boards.updateItem` same — doc `{ fields: { status } }`, canonical
 *     `{ data: [{ key, value }] }`.
 *   - `boards.createField type: "number"` lowercase — backend wants `Number`
 *     (capitalised). Doc shows lowercase.
 */
import "dotenv/config"
import { ImbraceClient } from "@imbrace/sdk"

const apiKey         = process.env.IMBRACE_API_KEY
const organizationId = process.env.IMBRACE_ORGANIZATION_ID
const baseUrl        = process.env.IMBRACE_GATEWAY_URL ?? "https://app-gatewayv2.imbrace.co"

if (!apiKey || !organizationId) {
  console.error("Missing IMBRACE_API_KEY or IMBRACE_ORGANIZATION_ID")
  process.exit(1)
}

const client = new ImbraceClient({ apiKey, organizationId, baseUrl, timeout: 15_000 })

let pass = 0, fail = 0, skip = 0
const fails: string[] = []
const docGaps: string[] = []

async function step(label: string, fn: () => Promise<unknown>, expectFail = false) {
  process.stdout.write(`  • ${label} ... `)
  try {
    const t0 = Date.now()
    const r = await fn()
    const dt = Date.now() - t0
    const summary = JSON.stringify(r ?? {}).slice(0, 90)
    if (expectFail) { console.log(`unexpected pass [${dt}ms]: ${summary}`); fail++; fails.push(`${label} → unexpected pass`) }
    else { console.log(`✓ [${dt}ms] ${summary}`); pass++ }
  } catch (err: any) {
    const code = err?.statusCode ?? err?.message ?? "ERR"
    if (expectFail) { console.log(`✓ (expected fail [${code}])`); pass++ }
    else { console.log(`✗ [${code}]`); fail++; fails.push(`${label} → ${code}`) }
  }
}

function skipped(label: string, reason: string) { console.log(`  - ${label}  ⏭ ${reason}`); skip++ }
function section(title: string) { console.log(`\n══ ${title} ══`) }
function note(msg: string) { console.log(`  ℹ ${msg}`); docGaps.push(msg) }

console.log(`\n━━━ DOCS: databoard.md — auth: API KEY (npm @imbrace/sdk@1.0.4) ━━━`)
console.log(`gateway=${baseUrl}\norg=${organizationId}\n`)

const ts = Date.now()
const state: { boardId: string | null; itemId: string | null; identifierFieldId: string | null; numberFieldId: string | null; segmentId: string | null } =
  { boardId: null, itemId: null, identifierFieldId: null, numberFieldId: null, segmentId: null }

// ── §1. Board CRUD ────────────────────────────────────────────────────────

section("§1. Board CRUD")

await step("boards.list", () => client.boards.list({ limit: 5 } as any))

await step("boards.create", async () => {
  const b: any = await client.boards.create({ name: `Enterprise Leads ${ts}` } as any)
  state.boardId = b?._id ?? b?.id ?? null
  return { id: state.boardId }
})

if (state.boardId) {
  await step("boards.get(boardId)", () => client.boards.get(state.boardId!))
  await step("boards.update(boardId, { name })", () =>
    client.boards.update(state.boardId!, { name: `Enterprise Leads 2025 ${ts}` } as any),
  )
} else {
  skipped("boards.get / boards.update", "no board fixture")
}

// ── §2. Items ─────────────────────────────────────────────────────────────

section("§2. Items (doc-gap: doc shows { fields: { name, status, value } } — not iterable)")
note("doc-gap: databoard.md §2 createItem shape `{ fields: { name, status, value } }` returns 400 — canonical is `{ fields: [{ board_field_id, value }] }`")
note("doc-gap: databoard.md §2 updateItem shape `{ fields: { status } }` returns 400 — canonical is `{ data: [{ key, value }] }`")

if (state.boardId) {
  await step("boards.listItems (limit 100, empty)",
    () => client.boards.listItems(state.boardId!, { limit: 100 } as any))

  // Need a field so we can populate the identifier
  await step("boards.createField (canonical Type=ShortText) — find identifier", async () => {
    const u: any = await client.boards.createField(state.boardId!, {
      name: "Company", type: "ShortText",
    } as any)
    const idf = (u?.fields ?? []).find((f: any) => f?.is_identifier)
    state.identifierFieldId = idf?._id ?? null
    return { fields: u?.fields?.length, identifierId: state.identifierFieldId }
  })

  // Doc-shape createItem (expected to fail 400)
  await step("boards.createItem (doc shape — expected fail)",
    () => client.boards.createItem(state.boardId!, { fields: { name: "Acme", status: "new", value: 50000 } } as any),
    /* expectFail */ true)

  // Canonical createItem
  if (state.identifierFieldId) {
    await step("boards.createItem (canonical shape)", async () => {
      const it: any = await client.boards.createItem(state.boardId!, {
        fields: [{ board_field_id: state.identifierFieldId!, value: `Acme Corp ${ts}` }],
      } as any)
      state.itemId = it?._id ?? it?.id ?? null
      return { id: state.itemId }
    })
  } else {
    skipped("boards.createItem (canonical)", "no identifier field")
  }

  if (state.itemId) {
    await step("boards.getItem", () => (client.boards as any).getItem(state.boardId!, state.itemId!))
    // Doc-shape updateItem (expected to fail 400)
    await step("boards.updateItem (doc shape — expected fail)",
      () => (client.boards as any).updateItem(state.boardId!, state.itemId!, { fields: { status: "qualified" } }),
      /* expectFail */ true)
    // Canonical updateItem
    if (state.identifierFieldId) {
      await step("boards.updateItem (canonical shape)", () =>
        (client.boards as any).updateItem(state.boardId!, state.itemId!, {
          data: [{ key: state.identifierFieldId!, value: `Acme Corp — qualified ${ts}` }],
        }),
      )
    }
    await step("boards.deleteItem", () => (client.boards as any).deleteItem(state.boardId!, state.itemId!))
    state.itemId = null
  } else {
    skipped("boards.getItem / updateItem / deleteItem", "no item fixture")
  }

  // bulkDeleteItems needs real ids; backend rejects empty list with 400.
  skipped("boards.bulkDeleteItems", "needs real item ids — backend rejects empty list with 400")
} else {
  skipped("§2 items lifecycle", "no board fixture")
}

// ── §3. Search ────────────────────────────────────────────────────────────

section("§3. Search")
if (state.boardId) {
  await step("boards.search (q='enterprise')",
    () => (client.boards as any).search(state.boardId!, { q: "enterprise" }))
} else {
  skipped("boards.search", "no board fixture")
}

// ── §4. Fields, Segments & Export ────────────────────────────────────────

section("§4. Fields, Segments & Export")
note("doc-gap: databoard.md §4 createField uses `type: \"number\"` (lowercase) — SDK 1.0.4 backend expects `\"Number\"` (capitalised). Same for `ShortText`, `LongText`, etc.")

if (state.boardId) {
  // Doc-shape lowercase type — likely fails 400
  await step("boards.createField (doc shape — type='number' lowercase)",
    () => client.boards.createField(state.boardId!, { name: "Deal Value", type: "number" } as any),
    /* expectFail */ true)

  // Canonical capitalised type
  await step("boards.createField (canonical — Number)", async () => {
    const f: any = await client.boards.createField(state.boardId!, { name: "Contract Value", type: "Number" } as any)
    const newField = (f?.fields ?? []).find((x: any) => x?.name === "Contract Value")
    state.numberFieldId = newField?._id ?? null
    return { id: state.numberFieldId }
  })

  if (state.numberFieldId) {
    await step("boards.updateField (rename)",
      () => (client.boards as any).updateField(state.boardId!, state.numberFieldId!, { name: "Contract Value v2" }))
    await step("boards.deleteField",
      () => (client.boards as any).deleteField(state.boardId!, state.numberFieldId!))
    state.numberFieldId = null
  } else {
    skipped("boards.updateField / deleteField", "no field fixture")
  }

  // Segments
  await step("boards.listSegments",
    () => (client.boards as any).listSegments(state.boardId!))
  await step("boards.createSegment", async () => {
    const s: any = await (client.boards as any).createSegment(state.boardId!, {
      name: `High Value Leads ${ts}`,
      filter: { field: "value", op: "gt", value: 10000 },
    })
    state.segmentId = s?._id ?? s?.id ?? null
    return { id: state.segmentId }
  })

  await step("boards.exportCsv",
    () => (client.boards as any).exportCsv(state.boardId!))
}

// ── Cleanup ───────────────────────────────────────────────────────────────

section("cleanup")
if (state.boardId) {
  await step("boards.delete (cleanup)", () => client.boards.delete(state.boardId!))
}

// ── Summary ───────────────────────────────────────────────────────────────

console.log(`\n━━━ Summary (databoard / API key) ━━━`)
console.log(`pass=${pass}  fail=${fail}  skip=${skip}`)
if (fails.length) { console.log("Failures:"); fails.forEach(f => console.log(`  - ${f}`)) }
if (docGaps.length) { console.log("Doc gaps to fix:"); docGaps.forEach(g => console.log(`  - ${g}`)) }
process.exit(fail > 0 ? 1 : 0)

File: test-accesstoken-pkg/py/test_docs_databoard.py
"""Mirrors website/public/sdk/databoard.md against `imbrace==1.0.4` (PyPI)
— Access Token auth. See test-api-pkg/py/test_docs_databoard.py for
doc-gap commentary.
"""
from __future__ import annotations
import json
import os
import sys
import time

from dotenv import load_dotenv
load_dotenv()

from imbrace import ImbraceClient

ACCESS_TOKEN = os.environ.get("IMBRACE_ACCESS_TOKEN")
ORG_ID       = os.environ.get("IMBRACE_ORGANIZATION_ID")
GATEWAY      = os.environ.get("IMBRACE_GATEWAY_URL", "https://app-gatewayv2.imbrace.co")

if not ACCESS_TOKEN or not ORG_ID:
    print("Missing IMBRACE_ACCESS_TOKEN or IMBRACE_ORGANIZATION_ID")
    sys.exit(1)

client = ImbraceClient(access_token=ACCESS_TOKEN, organization_id=ORG_ID, gateway=GATEWAY, timeout=15)

passed = failed = skipped_n = 0
failures: list[str] = []
doc_gaps: list[str] = []


def step(label, fn, expect_fail=False):
    global passed, failed
    sys.stdout.write(f"  - {label} ... "); sys.stdout.flush()
    try:
        t0 = time.time()
        r = fn()
        dt = int((time.time() - t0) * 1000)
        summary = json.dumps(r, default=str)[:90] if r is not None else ""
        if expect_fail:
            print(f"unexpected pass [{dt}ms] {summary}")
            failed += 1; failures.append(f"{label} -> unexpected pass")
        else:
            print(f"OK [{dt}ms] {summary}"); passed += 1
    except Exception as e:
        code = str(e)[:80]
        if expect_fail:
            print(f"OK (expected fail [{code}])"); passed += 1
        else:
            print(f"FAIL [{code}]"); failed += 1
            failures.append(f"{label} -> {code}")


def skip(label, reason):
    global skipped_n
    print(f"  - {label}  SKIP: {reason}"); skipped_n += 1


def section(title): print(f"\n== {title} ==")
def note(msg): print(f"  i {msg}"); doc_gaps.append(msg)


print(f"\n=== DOCS: databoard.md - auth: ACCESS TOKEN (PyPI imbrace==1.0.4) ===")
print(f"gateway={GATEWAY}\norg={ORG_ID}\n")

ts = int(time.time() * 1000)
state = {"boardId": None, "itemId": None, "identifierFieldId": None, "numberFieldId": None, "segmentId": None}


section("§1. Board CRUD")
note("doc-gap: databoard.md §1 Py shows `client.boards.create(\"Enterprise Leads\")` (positional str); SDK 1.0.4 takes keyword args (`name=..., description=...`)")

step("boards.list", lambda: client.boards.list(limit=5))


def _create_board():
    b = client.boards.create(name=f"Enterprise Leads {ts}")
    state["boardId"] = b.get("_id") if isinstance(b, dict) else getattr(b, "_id", None)
    return {"id": state["boardId"]}


step("boards.create (canonical: name=...)", _create_board)
if state["boardId"]:
    step("boards.get(board_id)", lambda: client.boards.get(state["boardId"]))
    step("boards.update(board_id, { name })",
         lambda: client.boards.update(state["boardId"], {"name": f"Enterprise Leads 2025 {ts}"}))
else:
    skip("boards.get / update", "no board fixture")


section("§2. Items (doc-gap: { fields: { ... } } shape — not iterable)")
note("doc-gap: databoard.md §2 create_item shape `{ \"fields\": { \"name\", \"status\", \"value\" } }` returns 400")
note("doc-gap: databoard.md §2 update_item shape `{ \"fields\": { \"status\" } }` returns 400")

if state["boardId"]:
    step("boards.list_items (limit 100, empty)",
         lambda: client.boards.list_items(state["boardId"], limit=100))


    def _create_field():
        u = client.boards.create_field(state["boardId"], {"name": "Company", "type": "ShortText"})
        fields = u.get("fields", []) if isinstance(u, dict) else []
        idf = next((f for f in fields if f.get("is_identifier")), None)
        state["identifierFieldId"] = idf.get("_id") if idf else None
        return {"fields": len(fields), "identifierId": state["identifierFieldId"]}

    step("boards.create_field (canonical Type=ShortText) — find identifier", _create_field)

    step("boards.create_item (doc shape — expected fail)",
         lambda: client.boards.create_item(state["boardId"], {"fields": {"name": "Acme", "status": "new", "value": 50000}}),
         expect_fail=True)

    if state["identifierFieldId"]:
        def _create_item():
            it = client.boards.create_item(state["boardId"], {
                "fields": [{"board_field_id": state["identifierFieldId"], "value": f"Acme Corp {ts}"}],
            })
            state["itemId"] = it.get("_id") if isinstance(it, dict) else getattr(it, "_id", None)
            return {"id": state["itemId"]}
        step("boards.create_item (canonical shape)", _create_item)
    else:
        skip("boards.create_item (canonical)", "no identifier field")

    if state["itemId"]:
        step("boards.get_item", lambda: client.boards.get_item(state["boardId"], state["itemId"]))
        step("boards.update_item (doc shape — expected fail)",
             lambda: client.boards.update_item(state["boardId"], state["itemId"], {"fields": {"status": "qualified"}}),
             expect_fail=True)
        if state["identifierFieldId"]:
            step("boards.update_item (canonical shape)",
                 lambda: client.boards.update_item(state["boardId"], state["itemId"], {
                     "data": [{"key": state["identifierFieldId"], "value": f"Acme Corp - qualified {ts}"}],
                 }))
        step("boards.delete_item", lambda: client.boards.delete_item(state["boardId"], state["itemId"]))
        state["itemId"] = None
    else:
        skip("boards.get_item / update_item / delete_item", "no item fixture")

    skip("boards.bulk_delete_items", "needs real item ids")
else:
    skip("§2 items lifecycle", "no board fixture")


section("§3. Search")
if state["boardId"]:
    step("boards.search (q='enterprise')", lambda: client.boards.search(state["boardId"], q="enterprise"))
else:
    skip("boards.search", "no board fixture")


section("§4. Fields, Segments & Export")
note("doc-gap: databoard.md §4 create_field uses `type: \"number\"` lowercase — SDK 1.0.4 backend expects `\"Number\"` capitalised")

if state["boardId"]:
    step("boards.create_field (doc shape — type='number' lowercase)",
         lambda: client.boards.create_field(state["boardId"], {"name": "Deal Value", "type": "number"}),
         expect_fail=True)


    def _create_number_field():
        f = client.boards.create_field(state["boardId"], {"name": "Contract Value", "type": "Number"})
        fields = f.get("fields", []) if isinstance(f, dict) else []
        new_field = next((x for x in fields if x.get("name") == "Contract Value"), None)
        state["numberFieldId"] = new_field.get("_id") if new_field else None
        return {"id": state["numberFieldId"]}

    step("boards.create_field (canonical — Number)", _create_number_field)

    if state["numberFieldId"]:
        step("boards.update_field (rename)",
             lambda: client.boards.update_field(state["boardId"], state["numberFieldId"], {"name": "Contract Value v2"}))
        step("boards.delete_field",
             lambda: client.boards.delete_field(state["boardId"], state["numberFieldId"]))
        state["numberFieldId"] = None
    else:
        skip("boards.update_field / delete_field", "no field fixture")

    step("boards.list_segments", lambda: client.boards.list_segments(state["boardId"]))


    def _create_segment():
        s = client.boards.create_segment(state["boardId"], {
            "name": f"High Value Leads {ts}",
            "filter": {"field": "value", "op": "gt", "value": 10000},
        })
        state["segmentId"] = s.get("_id") if isinstance(s, dict) else getattr(s, "_id", None)
        return {"id": state["segmentId"]}

    step("boards.create_segment", _create_segment)
    step("boards.export_csv", lambda: client.boards.export_csv(state["boardId"]))


section("cleanup")
if state["boardId"]:
    step("boards.delete (cleanup)", lambda: client.boards.delete(state["boardId"]))


print(f"\n=== Summary (databoard / Access Token) ===")
print(f"pass={passed}  fail={failed}  skip={skipped_n}")
if failures:
    print("Failures:")
    for f in failures: print(f"  - {f}")
if doc_gaps:
    print("Doc gaps to fix:")
    for g in doc_gaps: print(f"  - {g}")
sys.exit(1 if failed > 0 else 0)
